app/src/main/java/com/eveningoutpost/dexdrip/Models/LibreBlock.java

```diff
@@ -129,29 +135,58 @@ public static LibreBlock getLatestForTrend() {
         return getLatestForTrend(JoH.tsl() - Constants.DAY_IN_MS, JoH.tsl() );
     }
 
+    static LibreBlock getFromCursor(Cursor cursor) {
+        LibreBlock libreBlock  = new LibreBlock();
+
+        libreBlock.timestamp = cursor.getLong(cursor.getColumnIndex("timestamp"));
+        libreBlock.byte_start = cursor.getLong(cursor.getColumnIndex("bytestart"));
+        libreBlock.byte_end = cursor.getInt(cursor.getColumnIndex("byteend"));
+        libreBlock.reference = cursor.getString(cursor.getColumnIndex("reference"));
+        libreBlock.blockbytes = cursor.getBlob(cursor.getColumnIndex("blockbytes"));
+        libreBlock.calculated_bg = cursor.getDouble(cursor.getColumnIndex("calculatedbg"));
+        libreBlock.uuid = cursor.getString(cursor.getColumnIndex("uuid"));
+        libreBlock.patchUid = cursor.getBlob(cursor.getColumnIndex("patchUid"));
+
+        libreBlock.patchInfo = cursor.getBlob(cursor.getColumnIndex("patchInfo"));
+        return libreBlock;
+    }
 
     public static LibreBlock getLatestForTrend(long start_time, long end_time) {
 
-        return new Select()
-                .from(LibreBlock.class)
-                .where("bytestart == 0")
-                .where("byteend >= 344")
-                .where("timestamp >= ?", start_time)
-                .where("timestamp <= ?", end_time)
-                .orderBy("timestamp desc")
-                .executeSingle();
+        SQLiteDatabase db = Cache.openDatabase();
+
+        // Using this syntex since there is no way to tell the DB which index to use.
+        // Using ActiveAndroid method would take up to 8 seconds to complete.
+        Cursor cursor= db.rawQuery("select * from libreblock  INDEXED BY  index_LibreBlock_timestamp "+
+                "where bytestart == 0 AND byteend == 344 OR byteend == 44 " +
+                "AND timestamp >= " + start_time + " AND timestamp <= " + end_time +
```

Comment:
Does the SQL query need brackets to encapsulate the `OR` clauses? 

Do you need to make sure you close the cursor in the event of exception or other unexpected code path? I'm not familiar enough with sqlite optimization to know why you need to specify the index manually. Does the order of the clauses have any impact? 

I am concerned overall about the performance impact of the per minute Libre trend stuff so it would be very good to get this as efficient as possible.

Reply:
Yes, the SQL needs brackets for the OR, I have missed this.

I think that the code as written should be fine, but moved this using a try block to make sure that it gets closed.
I have tried changing the order of clauses but it did not help.

As for performance impact, there is nothing new here. Just adding the libre bluetooth. This changes will make things much faster for old code as well. In any case this feature is optional.

By the way, I believe that we are having the same issue on many of our DB queries. See for example: https://github.com/NightscoutFoundation/xDrip/issues/626